src/Sylius/Bundle/ResourceBundle/Storage/CookieStorage.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ResourceBundle\Storage;
+
+use Sylius\Component\Resource\Storage\StorageInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Cookie;
+use Symfony\Component\HttpFoundation\ParameterBag;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+final class CookieStorage implements StorageInterface, EventSubscriberInterface
```

Comment:
Cookie storage as a separate class fails at being cookie storage and it's an array storage (using only response cookies). There isn't any reason to replace the storage part with another implementation. 

If you want to use an external listener, you should add some additional behaviours to the storage itself, making it something more than regular `StorageInterface`:
 - populating the request cookies from the request
 - getting only the response cookies

And that's where implementation is leaking, how would you change storages then? Creating `CookieStorageInterface` extending `StorageInterface` and specifying the communication for that particular reason gives us no advantage. 

The current implementation is self-contained and has *one* responsibility - managing cookies. Feel free to contribute and propose your changes as code, not plain words.